### ISSUE TITLE
EAM force performance very poor with s=0

### ISSUE THREAD

**Issue by jameskermode** [JuliaMolSim/JuLIP.jl#139]
EAM force evaluation is very slow with spline accuracy `s=0`, ~1000x times slower than any other value of `s`. This is important because anything much looser than `s=0` doesn't match ASE/LAMMPS forces. Error is exacerbated for more sensitive quantities such as elastic constants and surface energies: for example `s=1e-9` gives 30% error in C44 and 100% error in surface energy.

Here's a script to show the problem (reason there's not a loop over `s` is that I get sporadic segfaults in the Julia garbage collector, which I don't want to get into right now; may also be associated with spline Fortran library, which is another reason to change it). 

Performance is comparable with JuLIP 0.10 and 0.12, i.e. before and after merging of #132, so it's not related to that.

```julia
using ASE
using JuLIP
using LinearAlgebra

@pyimport ase.calculators.eam as eam
atoms = bulk(:W) * [10, 10, 1]
rattle!(atoms, 0.1)

# datadep doesn't seem to work outside package; run `Pkg.test("JuLIP")` to download
#test_pots = joinpath(datadep"JuLIP_testpots", "JuLIP_data") * "/"

test_pots = homedir() * "/.julia/datadeps/JuLIP_testpots/JuLIP_data/"
W = test_pots * "w_eam4.fs"
ase_calc = ASECalculator(eam.EAM(potential=W))
f_ase = forces(ase_calc, atoms)

# warm up
julip_calc = EAM(W)
forces(julip_calc, atoms)

s = parse(Float64, ARGS[1])
julip_calc = EAM(W, s=s)
t = @elapsed f_julip = forces(julip_calc, atoms)
@show s, t, norm(f_julip - f_ase)
```

My output:
```shell
$ julia eam-perf.jl 1e-3
(s, t, norm(f_julip - f_ase)) = (0.001, 0.001885567, 0.5294790022588944)

$ julia eam-perf.jl 1e-6
(s, t, norm(f_julip - f_ase)) = (1.0e-6, 0.002470576, 0.09205453027120508)

$ julia eam-perf.jl 1e-9
(s, t, norm(f_julip - f_ase)) = (1.0e-9, 0.003440698, 0.003099750104086119)

$ julia eam-perf.jl 0.0
(s, t, norm(f_julip - f_ase)) = (0.0, 2.55706567, 6.894794381512263e-10)
```

Adding `@profile` to the `forces()` call shows ca. 75% of the time is spent in `Dierkx.derivative()` -- output pasted below, so I think we need to change the spline library. Based on discussion in #132, would Interpolations.jl be the thing to try?

```
$ julia --project=. ../eam-perf.jl 0.0
 Count  Overhead File                           Line Function
 =====  ======== ====                           ==== ========
     1         1 @Base/math.jl                   583 sqrt
     1         0 @StaticArrays/src/linalg.jl     226 macro expansion
     1         0 @StaticArrays/src/linalg.jl     214 _norm
     1         0 @StaticArrays/src/linalg.jl     213 norm
     1         0 @JuLIP/src/eam.jl                68 evaluate_d!(dEs::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{...
     4         0 @Base/reduce.jl                 242 mapreduce_impl(f::ComposedFunction{JuLIP.Potentials.SplinePairPotential, typeof(norm)}, op::type...
    11         0 @JuLIP/src/eam.jl                66 evaluate_d!(dEs::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{...
   189         0 @Base/reduce.jl                 245 macro expansion
   189         0 @Base/simdloop.jl                77 macro expansion
   189         0 @Base/reduce.jl                 243 mapreduce_impl(f::ComposedFunction{JuLIP.Potentials.SplinePairPotential, typeof(norm)}, op::type...
   193         0 @Base/operators.jl              938 ComposedFunction
   193         0 @Base/reduce.jl                 257 mapreduce_impl
   193         0 @Base/reduce.jl                 415 _mapreduce(f::ComposedFunction{JuLIP.Potentials.SplinePairPotential, typeof(norm)}, op::typeof(B...
   193         0 @Base/reducedim.jl              318 _mapreduce_dim
   193         0 @Base/reducedim.jl              310 #mapreduce#672
   193         0 @Base/reducedim.jl              310 mapreduce
   193         0 @Base/reducedim.jl              878 #_sum#682
   193         0 @Base/reducedim.jl              878 _sum
   193         0 @Base/reducedim.jl              874 #sum#680
   193         0 @Base/reducedim.jl              874 sum
   193         0 @JuLIP/src/eam.jl                65 evaluate_d!(dEs::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{...
   376       376 @Dierckx/src/Dierckx.jl         279 _evaluate(t::Vector{Float64}, c::Vector{Float64}, k::Int64, x::Float64, bc::Int64)
   376         0 @Dierckx/src/Dierckx.jl         296 evaluate
   376         0 @Dierckx/src/Dierckx.jl        1040 Spline1D
   376         0 @JuLIP/src/splines.jl            50 _evalspl
   376         0 @JuLIP/src/splines.jl            45 evaluate
   376         0 @JuLIP/src/potentials_base.jl    55 SplinePairPotential
   896         0 @JuLIP/src/pairpotentials.jl    240 evaluate_d
   896         0 @JuLIP/src/potentials_base.jl    56 SimpleProdPot
  1454      1454 @Dierckx/src/Dierckx.jl         324 _derivative(t::Vector{Float64}, c::Vector{Float64}, k::Int64, x::Float64, nu::Int64, bc::Int64, ...
  1454         0 @Dierckx/src/Dierckx.jl         344 derivative
  1454         0 @JuLIP/src/splines.jl            51 _evalspl_d
  1454         0 @JuLIP/src/splines.jl            46 evaluate_d
  1454         0 @JuLIP/src/potentials_base.jl    56 SplinePairPotential
  1626         0 @JuLIP/src/eam.jl                69 evaluate_d!(dEs::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{...
  1831         0 @JuLIP/src/Potentials.jl        114 evaluate_d!
  1831         0 @JuLIP/src/Potentials.jl        226 forces!(frc::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{Vect...
  1831         0 @JuLIP/src/Potentials.jl        220 forces!
  1831         0 @JuLIP/src/Potentials.jl        201 forces(V::JuLIP.Potentials.EAM1{JuLIP.Potentials.SimpleProdPot{JuLIP.Potentials.SplinePairPotent...
  1831         0 @JuLIP/src/Potentials.jl        201 forces(V::JuLIP.Potentials.EAM1{JuLIP.Potentials.SimpleProdPot{JuLIP.Potentials.SplinePairPotent...
  1833         2 @Base/boot.jl                   360 eval
  1833         0 @Base/loading.jl               1094 include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
  1833         0 @Base/loading.jl               1148 _include(mapexpr::Function, mod::Module, _path::String)
  1833         0 @Base/Base.jl                   386 include(mod::Module, _path::String)
  1833         0 @Base/client.jl                 285 exec_options(opts::Base.JLOptions)
  1833         0 @Base/client.jl                 485 _start()
Total snapshots: 2150
(s, t, norm(f_julip - f_ase)) = (0.0, 2.577403496, 7.388714445406024e-10)



[u1470235@lestrade tmp]$ julia --project=. ../eam-perf.jl 1e-3
 Count  Overhead File                           Line Function
 =====  ======== ====                           ==== ========
     1         1 @Dierckx/src/Dierckx.jl         324 _derivative(t::Vector{Float64}, c::Vector{Float64}, k::Int64, x::Float64, nu::Int64, bc::Int64, ...
     1         0 @Dierckx/src/Dierckx.jl         344 derivative
     1         0 @JuLIP/src/splines.jl            51 _evalspl_d
     1         0 @JuLIP/src/splines.jl            46 evaluate_d
     1         0 @JuLIP/src/potentials_base.jl    56 SplinePairPotential
     1         0 @JuLIP/src/eam.jl                69 evaluate_d!(dEs::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{...
     1         0 @JuLIP/src/Potentials.jl        114 evaluate_d!
     1         0 @JuLIP/src/Potentials.jl        226 forces!(frc::Vector{StaticArrays.SVector{3, Float64}}, tmp::NamedTuple{(:dV, :R, :Z), Tuple{Vect...
     1         0 @JuLIP/src/Potentials.jl        220 forces!
     1         0 @JuLIP/src/Potentials.jl        201 forces(V::JuLIP.Potentials.EAM1{JuLIP.Potentials.SimpleProdPot{JuLIP.Potentials.SplinePairPotent...
     1         0 @JuLIP/src/Potentials.jl        201 forces(V::JuLIP.Potentials.EAM1{JuLIP.Potentials.SimpleProdPot{JuLIP.Potentials.SplinePairPotent...
     3         2 @Base/boot.jl                   360 eval
     3         0 @Base/loading.jl               1094 include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
     3         0 @Base/loading.jl               1148 _include(mapexpr::Function, mod::Module, _path::String)
     3         0 @Base/Base.jl                   386 include(mod::Module, _path::String)
     3         0 @Base/client.jl                 285 exec_options(opts::Base.JLOptions)
     3         0 @Base/client.jl                 485 _start()
Total snapshots: 3
(s, t, norm(f_julip - f_ase)) = (0.001, 0.001853188, 0.5230623440778374)
```

**Comment by jameskermode** [comment]
I don't understand what the "overhead" column really means in the Profile output, and I see I'm not the only one: https://discourse.julialang.org/t/meaning-of-overhead-column-in-profile/51648, https://discourse.julialang.org/t/memory-allocation-and-profile/47573. I guess it's related to memory allocation/GC within Dierckx.

**Comment by jameskermode** [comment]
Here's the problematic line in Dierckx.jl: it's the `ccall` to the Fortran library. I guess argument conversion must be allocating?

https://github.com/kbarbary/Dierckx.jl/blob/master/src/Dierckx.jl#L324

**Comment by cortner** [comment]
That allocation problem has been fixed. But maybe I forgot to tag. 

Re the performance — thanks for documenting this. Obviously this is bad.

**Comment by cortner** [comment]
Initial thoughts : is there any point to in ever allowing a loose fit accuracy or will we always demand interpolation on the entire (huge!!!) nodal value set?

the main reason I haven’t dropped Dierckx is that is also has a regression code. But if we don’t need that then it should be easy to switch

**Comment by jameskermode** [comment]
Do you mean the allocation problem has been fixed within JuLIP? I'll try with latest `main`, I thought I'd done that but maybe not.

Re: fit accuracy, I would be happy to lose a bit of accuracy, but my tests show that properties are very sensitive to the spline accuracy.

**Comment by jameskermode** [comment]
The less accurate splines are not equivalent to just a slightly "different" model material either: for example we only get the expected rate of decay of forces ahead of a crack with `s=0`.

**Comment by jameskermode** [comment]
> I'll try with latest main, I thought I'd done that but maybe not.

Indeed, I'm on 0.12.3 == tip of `master`.

**Comment by mbuze** [comment]
> The less accurate splines are not equivalent to just a slightly "different" model material either: for example we only get the expected rate of decay of forces ahead of a crack with `s=0`.

Just to perhaps expand a bit on that: thanks to efforts by James we now have a Julia implementation of a general anisotropic CLE Mode I crack solution, all it needs is to be fed it with the three cubic elastic constants.

Regardless of the choice of `s`, we can compute elastic constants from the resulting model. We then displace the atoms according to the CLE solution obtained from the computed elastic constants and check the decay of forces. For anything other than `s=0` we see r^(-3/2), which seems to be implying that the elastic constants are wrong for this model (even though they were obtained from the model itself). On the other hand when `s=0`, we get the expected r^(-2) decay.

**Comment by cortner** [comment]
That’s just bizarre….

anyhow my question was unrelated to accuracy of the model. It was simply whether we will ever want something other than exact interpolation on the dataset. 

My guess is not so best to retire Dierckx

**Comment by jameskermode** [comment]
OK, I delved a bit deeper into Dierckx and understand better now - as you say, for `s=0` it simply retains all the knot points in the dataset.

```
julia> julip_calc = EAM(W, s=0);
julia> julip_calc.F[1].spl
Spline1D(knots=[0.0,0.000334193 … 16.7093,16.7097] (99999 elements), k=3, extrapolation="zero", residual=0.0)

julia> julip_calc = EAM(W, s=1e-12);
julia> julip_calc.F[1].spl
Spline1D(knots=[0.0,0.000334193 … 15.6653,16.7097] (135 elements), k=3, extrapolation="zero", residual=9.995608244276521e-13)
```

what I don't understand yet is how such small errors in the splines (here residual of 1e-12) mess up the potential so badly.

**Comment by jameskermode** [comment]
~100 knots are insufficient, and 10,000 are too slow. I’ll have to look how this is implemented in ASE and LAMMPS but I think they must also retain all spline points, just with a more efficient spline code.

**Comment by jameskermode** [comment]
Ah, I think it’s just the function call overhead that’s the problem, since Dierckx derivative is called once for each individual distance. Evaluating derivatives for a long vector of distances would be fine. Not sure how convenient it would be to reorganise things to do that, though. Will have a look later.

**Comment by cortner** [comment]
Unless this is really urgent don’t waste tour time. I’ve done this. Not sure sure where it is but I can find it again

**Comment by jameskermode** [comment]
Not urgent, happy to leave it for now.

**Comment by cortner** [comment]
ok, I have a simple solution - I'll use an Object pool for temporaries. But need to pause and think how/where to put it in. Please start pestering me if you need this.

**Comment by jameskermode** [comment]
My impression from profiling was that it was the function call overhead into Fortran that was the problem, not just the allocations, so I'm not sure this would be a full solution. I was trying to figure out how to vectorize the calls to spline evaluation and derivatives by collecting distances needed over all bonds, but also couldn't see any easy place to put this in without breaking the current encapsulation.

**Comment by cortner** [comment]
sorry for the delay; I've finally had time to test this. And you are right. This has nothing to do with temporaries ... as I said above I had already fixed the temporaries :) 

So the issue remains to either re-organize how the spline library is called or switch to a different library. 

Here is an even more radical proposal: why not split off the EAM code into EAM.jl? and then switch the spline library?

**Comment by cortner** [comment]
Some good news:  cost of a single spline evaluation 
```
[ Info: Dierckx: eval, grad
  2.618 μs (0 allocations: 0 bytes)
  19.461 μs (0 allocations: 0 bytes)
[ Info: Interpolations, eval, grad
  16.321 ns (0 allocations: 0 bytes)
  22.051 ns (0 allocations: 0 bytes)
```
Bad news : this only works with uniform knots. Now LAMMPS provides EAM files like this: 
```
 0.0
 0.0
 0.0
 0.0
 0.0012
 0.0018
 0.0024
 0.003
 0.0036
 0.0042
 0.0048
 0.0054
 0.006
....
```
so the first few knots are the same and then they become uniform. Same at the other end-point. Not sure yet how to hack this.

**Comment by gabor1** [comment]
Maybe lammps accepts arbitrary knot positions? Then you can make them uniform

-- Gábor

**Comment by cortner** [comment]
Here is a possible hack - but I'll wait for James' feedback before I implement it because it relies on this very specific structure I've printed above: 
* remove the first few nodes, [0.0,0.0,0.0,0.0] add one artificial node at the correct spacing so that it becomes 
```
[0.006, 0.0012, 0.0018, ...]
```
* replace the numerical knots with a `range` to tell the spline code that this is really uniform. 
* do the same at the other end 
* the values at the end points can be taken from the next ones in (1e12 and 0) or something like that 

this gives the following: slightly reduced performance (but still more than good enough!) 
```
[ Info: Dierckx: eval, grad
  3.125 μs (0 allocations: 0 bytes)
  19.973 μs (0 allocations: 0 bytes)
[ Info: Interpolations, eval, grad
  44.281 ns (0 allocations: 0 bytes)
  53.353 ns (0 allocations: 0 bytes)
```
and correctness tests all pass: 
```
[ Info: test for correctness
[ Info:   [1] random evaluation
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++[ Info: test for correctness
[ Info:   [2] close to cutoff
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++[ Info: test for correctness
[ Info:   [3] even close to cutoff
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
[ Info: test for correctness
[ Info:   [1] random evaluation
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
[ Info: test for correctness
[ Info:   [2] close to cutoff ~ 1e-4 
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
[ Info: test for correctness
[ Info:   [3] even closer to cutoff ~ 1e-12 
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

**Comment by cortner** [comment]
> Maybe lammps accepts arbitrary knot positions? Then you can make them uniform

Our problem is not to have lamps accept our knots, but to correctly represent the lammps splines in JuLIP. The issue is that they appear to be uniform which is really important for performance, otherwise the code has to search for the correct interval where to evaluate the splines. 

LAMMPS splines are indeed uniform except for some weird crap at the end-points. My suggestion above is a hack that restores uniformity and **APPEARS** to be correct on a first test case.

**Comment by jameskermode** [comment]
Thanks, sounds very sensible - I’d like to test this with some derived properties as I found they can depend very sensitively on spline knots. Can you share/push the modified code or shall I just implement based on your description?

**Comment by cortner** [comment]
I’ll implement a draft later and create a PR which you can then use to test

**Comment by jameskermode** [comment]
Perfect, thank you. I’ll also check in some of the EAM papers if there is an explanation for the duplicated knots.

**Comment by jameskermode** [comment]
https://openkim.org/publications/wen-msmse-2015.pdf looks relevant from a quick scan. Will look in more detail tomorrow.

**Comment by cortner** [comment]
closed by #140
